server/src/main/java/org/elasticsearch/index/query/functionscore/ScriptScoreQueryBuilder.java

```diff
@@ -0,0 +1,230 @@
+        }
+        if (scriptScoreFunctionBuilder == null) {
+            throw new IllegalArgumentException("script_score: script must not be null");
+        }
+        this.query = query;
+        this.scriptScoreFunctionBuilder = scriptScoreFunctionBuilder;
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public ScriptScoreQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        query = in.readNamedWriteable(QueryBuilder.class);
+        scriptScoreFunctionBuilder = in.readNamedWriteable(ScriptScoreFunctionBuilder.class);
+        minScore = in.readOptionalFloat();
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeNamedWriteable(query);
+        out.writeNamedWriteable(scriptScoreFunctionBuilder);
+        out.writeOptionalFloat(minScore);
+    }
+
+    /**
+     * Returns the query builder that defines which documents the script_score query will be executed on.
+     */
+    public QueryBuilder query() {
+        return this.query;
+    }
+
+    /**
+     * Returns the script function builder
+     */
+    public ScriptScoreFunctionBuilder scriptScoreFunctionBuilder() {
+        return this.scriptScoreFunctionBuilder;
+    }
+
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        if (query != null) {
+            builder.field(QUERY_FIELD.getPreferredName());
+            query.toXContent(builder, params);
+        }
+        builder.field(SCRIPT_FIELD.getPreferredName(), scriptScoreFunctionBuilder.getScript());
+        if (minScore != null) {
+            builder.field(MIN_SCORE_FIELD.getPreferredName(), minScore);
+        }
+        printBoostAndQueryName(builder);
+        builder.endObject();
+    }
+
+    public ScriptScoreQueryBuilder setMinScore(float minScore) {
+        this.minScore = minScore;
+        return this;
+    }
+
+    public Float getMinScore() {
+        return this.minScore;
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    protected boolean doEquals(ScriptScoreQueryBuilder other) {
+        return Objects.equals(this.query, other.query) &&
+            Objects.equals(this.scriptScoreFunctionBuilder, other.scriptScoreFunctionBuilder) &&
+            Objects.equals(this.minScore, other.minScore) ;
+    }
+
+    @Override
+    protected int doHashCode() {
+        return Objects.hash(this.query, this.scriptScoreFunctionBuilder, this.minScore);
+    }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        ScriptScoreFunction function = (ScriptScoreFunction) scriptScoreFunctionBuilder.toFunction(context);
+        Query query = this.query.toQuery(context);
+        if (query == null) {
+            query = new MatchAllDocsQuery();
```

Comment:
What conditions lead to the `toQuery()` call emitting `null`? I am concerned that assuming `null` can be converted to a `MatchAllDocsQuery` might be trappy here